app/services/fact_cleaner.rb

```diff
@@ -0,0 +1,44 @@
+class FactCleaner
+  delegate :logger, :to => :Rails
+
+  def clean!
+    @to_delete = []
+    logger.info "Searching for orphaned facts..."
+    find_leaf_orphaned_facts
+
+    # we also delete all composes that either don't have any descendant or all of them are to be deleted
+    logger.info "Searching for compose orphaned facts..."
+    find_compose_orphaned_facts
+
+    logger.info "#{@to_delete.size} facts will be removed"
+
+    delete!
+    logger.info "Cleanup of facts finished!"
+    @to_delete.count
+  end
+
+  private
+
+  def find_leaf_orphaned_facts
+    FactName.where(:compose => false).includes(:fact_values).find_in_batches do |batch|
+      @to_delete += batch.select do |fact|
+        fact.fact_values.empty?
+      end
+    end
+  end
+
+  def find_compose_orphaned_facts
+    FactName.where(:compose => true).find_in_batches do |batch|
+      @to_delete += batch.select do |compose|
+        compose.descendants.all? { |fact| @to_delete.include?(fact) }
+      end
+    end
+  end
+
+  def delete!
+    @to_delete.each do |fact|
```

Comment:
When users run Foreman they usually don't have sql logger enabled, for me it's easier to tell them how to enable debug log and rerun the task to see which fact failed than how to add sql logger, also it's more prone to errors since whitespaces in yaml matter. `destroy_all` does not bring much value, it instantiates every record and run all callback anyway.